### ISSUE TITLE
Unable to use BlenderBot2 model

### ISSUE THREAD

**Issue by LiShaoyu5** [facebookresearch/ParlAI#3808]
I tried to use pretrained blenderbot2 model from https://parl.ai/projects/blenderbot2/ but got error. Command and error message are as below:

`parlai interactive --model-file zoo:blenderbot2/blenderbot2_400M/model`

`Traceback (most recent call last):
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/build_data.py", line 490, in modelzoo_path
    my_module = importlib.import_module(module_name)
  File "/root/miniconda3/envs/myconda/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'parlai.zoo.blenderbot2'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/build_data.py", line 497, in modelzoo_path
    my_module = importlib.import_module(module_name_)
  File "/root/miniconda3/envs/myconda/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'parlai.zoo.blenderbot2'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/root/miniconda3/envs/myconda/bin/parlai", line 8, in <module>
    sys.exit(main())
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/__main__.py", line 14, in main
    superscript_main()
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/script.py", line 316, in superscript_main
    opt = parser.parse_args(args)
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/params.py", line 1163, in parse_args
    self.add_extra_args(args)
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/script.py", line 175, in add_extra_args
    sa.choices[args[0]].add_extra_args(args)
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/params.py", line 946, in add_extra_args
    model = get_model_name(parsed)
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/params.py", line 136, in get_model_name
    model_file = modelzoo_path(opt.get('datapath'), model_file)
  File "/root/miniconda3/envs/myconda/lib/python3.7/site-packages/parlai/core/build_data.py", line 504, in modelzoo_path
    ) from exc
ImportError: Could not find pretrained model in parlai.zoo.blenderbot2.blenderbot2_400M or parlai.zoo.blenderbot2.build. Please check your spelling and make sure you've pulled from master.`

**Comment by ibrahimjelliti** [comment]
same issue here, I tried to run both commands from [Wizard Of Internet models](https://parl.ai/docs/zoo.html?#wizard-of-internet-models)
 
`parlai interactive -mf zoo:blenderbot2/blenderbot2_400M/model --search-server relevant_search_server`
`parlai interactive -mf zoo:blenderbot2/blenderbot2_3B/model --search-server relevant_search_server
`

**Comment by stephenroller** [comment]
Try updating to the master branch. We didn't put out a pip release yet.

**Comment by ibrahimjelliti** [comment]
I am already up to date on master: i tried with this line, but same error model not found:
`python parlai/scripts/interactive.py -mf zoo:blenderbot2/blenderbot2_400M/model --search-server relevant_search_server`

**Comment by stephenroller** [comment]
Okay, the team will look on Monday. They're taking a very well earned break for now.

**Comment by ratthachat** [comment]
Similar issue with Colab
https://colab.research.google.com/drive/10hoBQbv_ArqfB3y7CcLRdQin5PbGOo78?usp=sharing
(few months ago, working fine on Blender Bot v 1.0)

**Comment by RodolfoFigueroa** [comment]
Are you sure you pulled master? It works fine on my colab environment (with the exception of search engine results, of course).
As a small note, I had to make some changes to the `requirements.txt`, because `pip` threw a fit otherwise:
```
huggingface-hub==0.0.12
requests<3,>=2.26.0
transformers==4.8.2
```

**Comment by ratthachat** [comment]
@RodolfoFigueroa Could you please share your working Colab ?

**Comment by RodolfoFigueroa** [comment]
> @RodolfoFigueroa Could you please share your working Colab ? (that would be great for the community)

Sure. https://colab.research.google.com/drive/1M-Pw6sQjR-0PtiouWL5LMEWWMrJnwurf?usp=sharing
I haven't managed to get the search queries to work, so for now it's limited to just training data. Also, you need a high-RAM environment to run the 3B model, change it to the 400M if you don't have Colab Pro.

**Comment by kazeto111** [comment]
Hi!  Thank you for your all great work for Blender2.0!!!! I was really waiting for this day all the time!
anyway I implemented 
`python setup.py develop`
`pip install -r requirement.txt`
`python parlai/scripts/interactive.py -mf zoo:blenderbot2/blenderbot2_400M/model --search-server relevant_search_server`

I didn't get the error like @LiShaoyu5 but got the one below

```
Traceback (most recent call last):
  File "/home/futo/ParlAI/parlai/core/build_data.py", line 490, in modelzoo_path
    my_module = importlib.import_module(module_name)
  File "/opt/conda/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'parlai.zoo.bart.bart_large'
```

I also followed the colab by @RodolfoFigueroa but the result was totally same.
Is this the same issue?
This can be conda-pip conflict? (I'm sorry if I'm looking in the wrong place.)

Thank you so much

**Comment by ibrahimjelliti** [comment]
According to internet-augmented Dialogue generation [README](https://github.com/facebookresearch/ParlAI/blob/master/projects/sea/README.md), search-server is a running search server that receives search queries and returns a set of documents. I think we need to setup this extra module first.

**Comment by ratthachat** [comment]
This is strange. I was able to run @RodolfoFigueroa colab once, but now I have the same problem as @kazeto111

**Comment by ibrahimjelliti** [comment]
400M model doesn't run for me on free google Colab

**Comment by stephenroller** [comment]
I was able to get the 400M to work on the colab just fine.

**Comment by klshuster** [comment]
Hi all - like @stephenroller, I was able to run the 400m model on colab just fine. Could you please paste some full stack traces for when it does not work for you?

The only issue I've run into so far is download issues if I don't leave colab as my main chrome screen

**Comment by ibrahimjelliti** [comment]
@klshuster 
> 20:51:05 | building data: /content/ParlAI/data/models/blenderbot2/blenderbot2_3B/model.tgz
20:51:05 | Downloading http://parl.ai/downloads/_models/blenderbot2/blenderbot2_3B/model.tgz to /content/ParlAI/data/models/blenderbot2/blenderbot2_3B/model.tgz
Downloading model.tgz: 100% 5.54G/5.54G [03:09<00:00, 29.3MB/s]
20:57:06 | Overriding opt["model_file"] to /content/ParlAI/data/models/blenderbot2/blenderbot2_3B/model (previously: zoo:blenderbot2/blenderbot2_3B/model)
20:57:06 | Overriding opt["search_server"] to http://asd (previously: None)
20:57:06 | loading dictionary from /content/ParlAI/data/models/blenderbot2/blenderbot2_3B/model.dict
20:57:06 | num words = 8008
20:57:06 | BlenderBot2Fid: full interactive mode on.
Downloading: 100% 232k/232k [00:00<00:00, 1.61MB/s]
Downloading: 100% 466k/466k [00:00<00:00, 2.76MB/s]
Downloading: 100% 28.0/28.0 [00:00<00:00, 21.0kB/s]
Downloading: 100% 570/570 [00:00<00:00, 424kB/s]
20:57:51 | building data: /content/ParlAI/data/models/hallucination/bart_rag_token/model.tgz
20:57:51 | Downloading http://parl.ai/downloads/_models/hallucination/bart_rag_token/model.tgz to /content/ParlAI/data/models/hallucination/bart_rag_token/model.tgz
Downloading model.tgz: 100% 950M/950M [00:31<00:00, 30.4MB/s]
20:58:43 | Creating the search engine retriever.
20:58:46 | building data: /content/ParlAI/data/models/hallucination/multiset_dpr/hf_bert_base.cp
20:58:46 | Downloading https://dl.fbaipublicfiles.com/dpr/checkpoint/retriver/multiset/hf_bert_base.cp to /content/ParlAI/data/models/hallucination/multiset_dpr/hf_bert_base.cp
Downloading hf_bert_base.cp: 100% 876M/876M [00:28<00:00, 31.1MB/s]
Downloading: 100% 440M/440M [00:07<00:00, 59.9MB/s]
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.dense.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.decoder.weight', 'cls.predictions.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
20:59:28 | building data: /content/ParlAI/data/models/blenderbot2/query_generator/model.tgz
20:59:28 | Downloading http://parl.ai/downloads/_models/blenderbot2/query_generator/model.tgz to /content/ParlAI/data/models/blenderbot2/query_generator/model.tgz
Downloading model.tgz: 100% 750M/750M [00:25<00:00, 29.9MB/s]
21:00:07 | Building Query Generator from file: /content/ParlAI/data/models/blenderbot2/query_generator/model
Downloading vocab.bpe: 0.00B [00:00, ?B/s]
Downloading encoder.json: 0.00B [00:00, ?B/s]
21:00:16 | building data: /content/ParlAI/data/models/blenderbot2/memory_decoder/model.tgz
21:00:16 | Downloading http://parl.ai/downloads/_models/blenderbot2/memory_decoder/model.tgz to /content/ParlAI/data/models/blenderbot2/memory_decoder/model.tgz
Downloading model.tgz: 100% 750M/750M [00:24<00:00, 30.6MB/s]
21:00:54 | Building Memory Decoder from file: /content/ParlAI/data/models/blenderbot2/memory_decoder/model
[downloading BART models: /content/ParlAI/data/models/bart]
Downloading bart.large.tar.gz:  65% 2.42G/3.70G [01:27<00:52, 24.2MB/s]Traceback (most recent call last):
  File "/content/ParlAI/parlai/core/build_data.py", line 490, in modelzoo_path
    my_module = importlib.import_module(module_name)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'parlai.zoo.bart.bart_large'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/urllib3/response.py", line 438, in _error_catcher
    yield
  File "/usr/local/lib/python3.7/dist-packages/urllib3/response.py", line 519, in read
    data = self._fp.read(amt) if not fp_closed else b""
  File "/usr/lib/python3.7/http/client.py", line 465, in read
    n = self.readinto(b)
  File "/usr/lib/python3.7/http/client.py", line 509, in readinto
    n = self.fp.readinto(b)
  File "/usr/lib/python3.7/socket.py", line 589, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/requests/models.py", line 758, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/usr/local/lib/python3.7/dist-packages/urllib3/response.py", line 576, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/local/lib/python3.7/dist-packages/urllib3/response.py", line 541, in read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
  File "/usr/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.7/dist-packages/urllib3/response.py", line 455, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
urllib3.exceptions.ProtocolError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/parlai", line 33, in <module>
    sys.exit(load_entry_point('parlai', 'console_scripts', 'parlai')())
  File "/content/ParlAI/parlai/__main__.py", line 14, in main
    superscript_main()
  File "/content/ParlAI/parlai/core/script.py", line 325, in superscript_main
    return SCRIPT_REGISTRY[cmd].klass._run_from_parser_and_opt(opt, parser)
  File "/content/ParlAI/parlai/core/script.py", line 108, in _run_from_parser_and_opt
    return script.run()
  File "/content/ParlAI/parlai/scripts/interactive.py", line 118, in run
    return interactive(self.opt)
  File "/content/ParlAI/parlai/scripts/interactive.py", line 84, in interactive
    agent = create_agent(opt, requireModelExists=True)
  File "/content/ParlAI/parlai/core/agents.py", line 468, in create_agent
    model = create_agent_from_opt_file(opt)
  File "/content/ParlAI/parlai/core/agents.py", line 421, in create_agent_from_opt_file
    return model_class(opt_from_file)
  File "/content/ParlAI/parlai/agents/rag/rag.py", line 174, in __init__
    self._generation_agent.__init__(self, opt, shared)  # type: ignore
  File "/content/ParlAI/parlai/core/torch_generator_agent.py", line 484, in __init__
    self.model = fsdp_utils.fsdp_wrap(self.build_model())
  File "/content/ParlAI/projects/blenderbot2/agents/blenderbot2.py", line 885, in build_model
    model = BlenderBot2FidModel(self.opt, self.dict)
  File "/content/ParlAI/projects/blenderbot2/agents/modules.py", line 97, in __init__
    self.memory_decoder = MemoryDecoder(opt)
  File "/content/ParlAI/projects/blenderbot2/agents/sub_modules.py", line 282, in __init__
    model_file, opt_overrides=overrides
  File "/content/ParlAI/parlai/core/agents.py", line 347, in create_agent_from_model_file
    return create_agent_from_opt_file(opt)
  File "/content/ParlAI/parlai/core/agents.py", line 421, in create_agent_from_opt_file
    return model_class(opt_from_file)
  File "/content/ParlAI/parlai/agents/bart/bart.py", line 71, in __init__
    opt = self._initialize_bart(opt)
  File "/content/ParlAI/parlai/agents/bart/bart.py", line 97, in _initialize_bart
    compare_init_model_opts(opt, opt)
  File "/content/ParlAI/parlai/core/agents.py", line 273, in compare_init_model_opts
    opt['init_model'] = modelzoo_path(opt['datapath'], opt['init_model'])
  File "/content/ParlAI/parlai/core/build_data.py", line 498, in modelzoo_path
    my_module.download(datapath)
  File "/content/ParlAI/parlai/zoo/bart/build.py", line 64, in download
    build_data.download(url, dpath, f'{model_name}.tar.gz')
  File "/content/ParlAI/parlai/core/build_data.py", line 181, in download
    for chunk in response.iter_content(CHUNK_SIZE):
  File "/usr/local/lib/python3.7/dist-packages/requests/models.py", line 761, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
Downloading bart.large.tar.gz:  65% 2.42G/3.70G [01:27<00:46, 27.5MB/s]

**Comment by klshuster** [comment]
@ibrahimjelliti yeah this appears to be the issue I was having as well. If you open up the files tab in the colab on the left side and to to `ParlAI/data/models/blenderbot2/`, i'd delete everything inside the `bart` folder (and the `bart` folder itself) and try again.

This model does require quite a few downloads (BART, query generator, memory decoder, long-term-memory retriever, and BB2 itself) and I noticed that colab was acting a bit finicky with the download process... I got several of those `ConnectionResetError` exceptions as well if I put the tab in the background.

**Comment by JulesGM** [comment]
You also all need to change the `relevant_search_server` from the command to a valid server.

There isn't one in the package, so I coded one:
https://github.com/JulesGM/ParlAI_SearchEngine

**Comment by kazeto111** [comment]
I followed @klshuster and it worked well!
Thank you so much!!!

**Comment by LiShaoyu5** [comment]
I installed parlai again in development mode and the problem was fixed. thank you!

**Comment by hitchingsh** [comment]
Shouldn't this git project also provide the search server?

**Comment by JulesGM** [comment]
They likely used ie Azure search services, so you would have to pay

**Comment by ratthachat** [comment]
@JulesGM Thanks so much. Is it possible to run your search engine code in colab?

**Comment by RodolfoFigueroa** [comment]
> Shouldn't this git project also provide the search server?

If you take a look at the code they seem to have used some custom search server, since they only pass two parameters to the request: the query and the number of results to return.

**Comment by JulesGM** [comment]
@ratthachat Yes. Run this in one of the cell before starting the bot:

```python
import multiprocess
import subprocess
PATH_TO_SEARCH_SERVER = "./search_server.py"
def start_server():
    subprocess.check_call(f"python {PATH_TO_SEARCH_SERVER} serve", shell=True)
multiprocess.Process(target=start_server).start()
```

modify PATH_TO_SEARCH_SERVER as needed.

**Comment by JulesGM** [comment]
added information on how to use the script with colab in the readme

**Comment by JulesGM** [comment]
Colab Update: Running the 400M .. runs out of memory. So it's not possible. Other people have reported this mistake.

**Comment by ratthachat** [comment]
Thanks @JulesGM , perhaps it could be run with High-RAM instance ?

**Comment by JulesGM** [comment]
@ratthachat  you can try https://github.com/JulesGM/ParlAI_SearchEngine/blob/master/colab.ipynb
some people have told me it worked for them in regular colab instances

**Comment by alelasantillan** [comment]
> @ratthachat you can try https://github.com/JulesGM/ParlAI_SearchEngine/blob/master/colab.ipynb
> some people have told me it worked for them in regular colab instances

Hi @JulesGM , great work! 
I was wondering in which way can I add to your ParlAI_SearchEngine a custom directory with text files so the relevant documents can be found also in that directory. This way one can add custom information to the web search.
Any idea?

Thank you!

**Comment by alelasantillan** [comment]
@JulesGM When running your colab notebook at https://github.com/JulesGM/ParlAI_SearchEngine/blob/master/colab.ipynb
last cell hangs as follows below, when downloading bart_large model. Any idea why?

!python -m parlai interactive --model-file zoo:blenderbot2/blenderbot2_400M/model --search_server $HOST
20:05:51 | Overriding opt["model_file"] to /usr/local/lib/python3.7/dist-packages/data/models/blenderbot2/blenderbot2_400M/model (previously: /checkpoint/kshuster/projects/knowledge_bot/kbot_memfix_sweep25_Fri_Jul__9/338/model.oss)
20:05:51 | Overriding opt["search_server"] to 0.0.0.0:1111 (previously: None)
20:05:51 | Using CUDA
20:05:51 | loading dictionary from /usr/local/lib/python3.7/dist-packages/data/models/blenderbot2/blenderbot2_400M/model.dict
20:05:51 | num words = 50264
20:05:51 | BlenderBot2Fid: full interactive mode on.
20:06:28 | Creating the search engine retriever.
20:06:28 | No portocol provided, using "http://"
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.dense.bias', 'cls.seq_relationship.weight', 'cls.predictions.decoder.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.bias']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
20:07:03 | Building Query Generator from file: /usr/local/lib/python3.7/dist-packages/data/models/blenderbot2/query_generator/model
20:07:40 | Building Memory Decoder from file: /usr/local/lib/python3.7/dist-packages/data/models/blenderbot2/memory_decoder/model
[downloading BART models: /usr/local/lib/python3.7/dist-packages/data/models/bart]
Downloading bart.large.tar.gz: 0.00B [00:00, ?B/s]
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/build_data.py", line 490, in modelzoo_path
    my_module = importlib.import_module(module_name)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'parlai.zoo.bart.bart_large'

**Comment by JulesGM** [comment]
are you sure you installed Fairseq?

**Comment by alelasantillan** [comment]
> are you sure you installed Fairseq?
> […](#)


Actually now works perfect. I think it was a colab thing. I did not change the code. I've just run it several times at different hours. 

I am still wondering in which way can I add to your ParlAI_SearchEngine a custom directory with text files so the relevant documents can be found also in that directory. This way one can add custom information to the web search.
Any idea?

**Comment by JulesGM** [comment]
@alelasantillan please open an issue on the server project page, commenting here pings a lot of other people.
